DDP-5806 File Upload extended functionality

alextanlan opened the pull request:
**_Update for 2 Jun, 2021:_**
![1](https://user-images.githubusercontent.com/7396837/120496151-0208bc80-c3c6-11eb-9bb8-38ec3f563ec7.png)
![1_1](https://user-images.githubusercontent.com/7396837/120496155-02a15300-c3c6-11eb-9f9b-b98837b46a06.png)
![2](https://user-images.githubusercontent.com/7396837/120496156-02a15300-c3c6-11eb-8e98-423b0a35aa7e.png)
![3](https://user-images.githubusercontent.com/7396837/120496158-0339e980-c3c6-11eb-9bbe-abc772af44c2.png)
![3_1](https://user-images.githubusercontent.com/7396837/120496160-0339e980-c3c6-11eb-98c8-5dc5e269c878.png)
![4](https://user-images.githubusercontent.com/7396837/120496161-0339e980-c3c6-11eb-8ecf-702207c65f16.png)
![4_1](https://user-images.githubusercontent.com/7396837/120496163-03d28000-c3c6-11eb-80b7-279892321296.png)
![image](https://user-images.githubusercontent.com/7396837/120705423-7aec3f00-c4c0-11eb-91dd-de845d10f4e6.png)

alextanlan committed:
Add '400 error' for answer failure handling

alextanlan committed:
Add deletion/replacement of uploaded file

alextanlan committed:
Merge branch 'develop' into DDP-5806-file-upload-extended

alextanlan committed:
Refactoring

  transform specific ActivityDeleteDialogComponent into common ConfirmDialog

alextanlan committed:
Apply confirm dialogs for fileUpload

alextanlan committed:
Add validation for fileAnswer

alextanlan committed:
Add tests file for fileAnswer

alextanlan committed:
Merge branch 'develop' into DDP-5806-file-upload-extended

# Conflicts:
#	ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json

alextanlan committed:
Add more tests for fileAnswer

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+    <div class="file-types" *ngIf="fileMimeTypes.length">Allowed mime-types: {{fileMimeTypes.join(', ')}}</div>
+</div>
+
+<div class="file-upload-content" [ngClass]="{'readonly': readonly}">
```
#minor there is a recomendation from Angular team 
https://angular.io/guide/built-in-directives#adding-and-removing-classes-with-ngclass
"To add or remove a single class, use class binding rather than NgClass."
means `[class.readonly]="readonly"`

alextanlan replied:
fixed at 4e8150675e0c6ee9724cd52ec02d8f73120cc52a

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.scss:
```diff
@@ -9,22 +30,23 @@
-.drop-block.file-over {
-    background-color: lightyellow;
-}
+    &.file-over {
```
wondering if we need to change the styles for that for readonly state

alextanlan replied:
It is overridden on line 13:

```
.file-upload-content.readonly {
    opacity: 0.7;

    .drop-block {
        background-color: $drop-block-color;
        cursor: not-allowed;
    }
```

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+        </button>
     </div>
-    <div class="file-info">
-        <span class="file-name">{{file?.name || 'No file uploaded'}}</span>
```
don't we need "No file uploaded" message anymore?

alextanlan replied:
There is no design mockup for that.
I leave an empty space if no uploaded file, for now.

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+    <div class="uploaded-file">
+        <div class="title">Uploaded:</div>
+
+        <mat-chip-list>
```
are we going to support multiple files uploading in the future? otherwise we probably can omit mat-chip-list

alextanlan replied:
There is no any design mockup for 'File Upload' functionality.
Not sure if we are going to support multiple files uploading.
I have chosen `mat-chip` because it is already an existing material component with included remove functionality.

aglushko-broad replied:
I see, optionally you can consider just remove icon-button next to the file name

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.spec.ts:
```diff
@@ -0,0 +1,203 @@
+
+        it('should display an uploaded file data', () => {
+            const uploadedFile = fixture.debugElement.query(By.css('.uploaded-file-chip')).nativeElement;
+            expect(uploadedFile.textContent.trim()).toBe('1.png (size: 1000) cancel');
```
#minor we can skip icon content using toContain instead toBe method

alextanlan replied:
Thanks, fixed at 4e8150675e0c6ee9724cd52ec02d8f73120cc52a

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.spec.ts:
```diff
@@ -0,0 +1,203 @@
+        it('should call undoUploadedFile by click on remove uploaded file button', () => {
+            spyOn(component, 'undoUploadedFile');
+            const removeUploadedFileBtn = fixture.debugElement.query(By.css('.uploaded-file mat-icon')).nativeElement;
+            expect(removeUploadedFileBtn).toBeTruthy();
```
#minor we can remove this check since if it's not truthy the test will throw error and fail anyway

alextanlan replied:
Fixed at 0f692defcc2f194893faa6d53ca7b38a2ac473fc

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.spec.ts:
```diff
@@ -0,0 +1,203 @@
+            expect(component.errorMessage).toBe(errorMessage);
+        });
+    });
+});
```
Cool!

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+        if (this.block?.answer) {
+            this.setUploadedFile({
+                uploadGuid: '',
+                uploadUrl: ',',
```
should it be just an empty string?

alextanlan replied:
Right, it is a typo, thanks.
fixed at 0f692defcc2f194893faa6d53ca7b38a2ac473fc

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+        }
+    }
+
+    public openConfirmDialog(): void {
```
should it be private?

alextanlan replied:
Technically yes, it could be.
However it is used in several test cases.

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+        }
+    }
+
+    public openConfirmDialog(): void {
```
can we rename to 'openReuploadConfirmDualog' to better specify the goal?

alextanlan replied:
Deal, fixed at 0f692defcc2f194893faa6d53ca7b38a2ac473fc

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+    }
+
+    undoUploadedFile(): void {
+        const panelClass = 'file-upload-confirm-dialog';
```
if this panel class is common for all dialogs of this component can we move it to private readonly class property?

alextanlan replied:
Fixed at 0f692defcc2f194893faa6d53ca7b38a2ac473fc

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
-    ngOnDestroy(): void {
-        this.ngUnsubscribe.next();
-        this.ngUnsubscribe.complete();
+    private setUploadedFile(file: UploadFile | null): void {
```
#minor I'm not sure we need this set method... are we going to extend it with more functionality? if not probably it's fine to use the explicit set in the code

alextanlan replied:
let's keep it as is

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+    }
+
+    private patchAnswer(file: UploadFile | null): void {
+        this.block.answer = file ? { fileName: file.name, fileSize: file.size } as ActivityFileAnswerDto : null;
```
why do we need type casting here?

alextanlan replied:
Just for more comprehension)
Removed at 4e8150675e0c6ee9724cd52ec02d8f73120cc52a

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -58,31 +81,130 @@ export class ActivityFileAnswer implements OnDestroy {
+            .find(validator => {
+                const preUploadFileAnswer: ActivityFileAnswerDto = {
+                    fileName: this.fileToUpload && this.fileToUpload.name,
+                    fileSize: this.fileToUpload && this.fileToUpload.size
```
#minor 
<!--StartFragment-->
fileName: this.fileToUpload?.name,                    
fileSize: this.fileToUpload?.size
<!--EndFragment-->

alextanlan replied:
Fixed at 4e8150675e0c6ee9724cd52ec02d8f73120cc52a, thanks.

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -1,51 +1,74 @@
+    fileMaxSize: number;
+    fileMimeTypes: string[];
+    errorMessage: string;
+    private fileToUpload: UploadFile | null;
```
it seems like it was better before with fileUploadData property. In fact, fileToUpload property is a combination of fileUploadData property and selectedFile property which means we duplicate information in the class instance state. I'd get rid of selectedFile then or get back to fileUploadData and just create the fileToUpload object locally in methods where it's needed.

alextanlan replied:
Refactored at 497c9ebd1345abb708c9fd0851e40d38fe7405b5

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/models/uploadFile.ts:
```diff
@@ -0,0 +1,7 @@
+import { FileUploadResponse } from './fileUploadResponse';
+
+export interface UploadFile extends FileUploadResponse {
```
I'm not sure if there is a benefit of putting the local component interfaces into the model folder. This model does not participate in contracts between backend and frontend and is not reused between different parts of the app. My suggetion is keeping such interfaces in the component which uses it. To keep the scope small enough.

alextanlan replied:
I don't think it is an issue. 
@MocanaAtBroad, what is your opinion ?

MocanaAtBroad replied:
I think it is OK to for it to be in the `models` directory as interface gets used from multiple places (component and service) we don't have a better home for it.
But @aglushko-broad observation about a distinction between classes and interfaces that are part of the client/server API vs those that are just for internal consumption is a good one and we should consider doing a little bit of refactoring here. (I did a quick count. We have 63 files directly in model directory, so probably time to reorganize the closet).

MocanaAtBroad replied:
After last refactor and reviewing latest changes, I am under impression that this interface can just go away. Only thing in it that matters is the File (one attribute) and the upload url is the other argument to actually do the upload and appears it would be easier to just pass it as a function argument. **_MIGHT_** be able to get rid of fileToUpload instance variable altogether, but not 100% sure, but please do take a look. The less instance variables we have, the less state to track and less chance of subtle bugs coming into play.

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/activity/validators/activityFileValidationRule.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private isMaxFileSizeExceeded(answer: ActivityFileAnswerDto): boolean {
+        return this.block.maxFileSize && answer && answer.fileSize > this.block.maxFileSize;
```
#minor answer && answer.fileSize => answer?.fileSize
same for the line 33

alextanlan replied:
Ok, fixed at 4e8150675e0c6ee9724cd52ec02d8f73120cc52a

alextanlan committed:
Add tests for file upload service

alextanlan committed:
Tweak validation

alextanlan committed:
Fix code-review comments

alextanlan committed:
Fix more code-review comments

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+
+<div class="file-requirements" *ngIf="fileMaxSize || fileMimeTypes.length">
+    <div class="file-max-size" *ngIf="fileMaxSize">Max size: {{fileMaxSize}}</div>
+    <div class="file-types" *ngIf="fileMimeTypes.length">Allowed mime-types: {{fileMimeTypes.join(', ')}}</div>
```
Strings need to be internationalized.

MocanaAtBroad replied:
Let's convert max size to Kilobytes or Megabytes using appropriate abbreviation.

MocanaAtBroad replied:
We probably need to refer to file types or file extensions. People don't like mimes.

alextanlan replied:
Added format:
- file size to Mb/kB
- mime-type to file extension
f49dcc17a39e5a7d0292664acf067ce18b74151a

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+                mat-raised-button color="primary"
+                class="submit-btn"
+                [disabled]="!isFileSelected"
+                (click)="submitFileUpload()">{{uploadedFile?.isUploaded? 'Reupload' : 'Submit'}}
```
Internationalization.

MocanaAtBroad replied:
We should give the button a label like "Upload" rather than Submit. We have been using "Submit" at the completion of forms.

alextanlan replied:
Renamed the label at 1e5f66ed5e3c0ab09f821ff006e5324212f4d37e

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
-        <span class="file-name">{{file?.name || 'No file uploaded'}}</span>
+
+    <div class="uploaded-file">
+        <div class="title">Uploaded:</div>
```
Same. Please check all code for strings that need to be i18n'ed.

alextanlan replied:
**_As for i18n:_**
I have added i18n for ddp-angio only (in test purpose) at 96f90878ac4a661c1aa6a55c2216e04e08aabeb3

we don't know for now where the File Upload functionality will be applied (in which activity).
It does not make sense to apply FileUpload translations for all existing studies right now, in my opinion. 

What do you think about it?

MocanaAtBroad replied:
Not needed in any of our current studies. But coming soon!

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/activity/validators/activityFileValidationRule.ts:
```diff
@@ -0,0 +1,36 @@
+        const answer = answerBeforeUpload || this.block.answer;
+
+        if (this.isMaxFileSizeExceeded(answer)) {
+            this.result = 'File size exceeded maximum of 100 bytes';
```
Internationalization and hard-coded value.

alextanlan replied:
Extracted hard-coded values at 9e99630b172eb4b407212bdcde6dc0fc3d1b35c7

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/activity/validators/activityFileValidationRule.ts:
```diff
@@ -0,0 +1,36 @@
+import { ActivityFileQuestionBlock } from '../../../models/activity/activityFileQuestionBlock';
+import { ActivityFileAnswerDto } from '../../../models/activity/activityFileAnswerDto';
+
+export class ActivityFileValidationRule extends ActivityAbstractValidationRule {
```
Actually, does this rule ever execute?
When trying add a large file, the displayed error appears to come from server.

alextanlan replied:
It's my pain)
Before the implementation I have been told that there is no any server pre-validation. 
Server only sends fields `maxFileSize`/ `allowed mime-types` as no notify client about the rules.
I created a local client validation using these fields via `ActivityFileValidationRule` and it executed.
Then (suddenly) Eugenii implemented server-side validation on getUploadUrl step.
And for now the client-side validation is kind of `backup validation`.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/activity/validators/activityFileValidationRule.ts:
```diff
@@ -0,0 +1,36 @@
+        super(block);
+    }
+
+    public recalculate(answerBeforeUpload?: ActivityFileAnswerDto): boolean {
```
If we are going to execute this rule or show the max size for file, let's display the file size in either KBs or MBs instead of just bytes. In Mac and Windows files in bytes are rarely seen, and we people should not have to do calculation.

alextanlan replied:
Added file size format pipe as to transform to Mb/kb and use in in template.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/services/modal.service.ts:
```diff
@@ -14,26 +14,26 @@ const DIALOG_HEIGHT = 160;
 
 @Injectable()
-export class ActivityBlockModalService {
+export class ModalService {
```
I think the refactor is a good idea.
Can we rename to ModalDialogService?

alextanlan replied:
Renamed at fdaba73aac86dc5241544432083394fae34022a1

MocanaAtBroad commented on ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:
```diff
@@ -699,6 +699,13 @@
+        },
+        "FileUpload": {
+            "ConfirmReuploadTitle": "Are you sure you want \nto reupload the file?",
+            "ConfirmUndoUploadTitle": "Are you sure you want \nto undo the uploaded file?",
```
Consider rewording to something like: "Are you sure you want to remove uploaded file from server?"

alextanlan replied:
Fixed at 9a8302427db3a7037cfa1c1715618308c8ff8e51

MocanaAtBroad commented on ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:
```diff
@@ -699,6 +699,13 @@
+            "ConfirmReuploadTitle": "Are you sure you want \nto reupload the file?",
+            "ConfirmUndoUploadTitle": "Are you sure you want \nto undo the uploaded file?",
+            "ReuploadBtnText": "Reupload",
+            "UndoBtnText": "Undo",
```
I think "OK" would be more familiar than "Undo" to users.

alextanlan replied:
Fixed at 9a8302427db3a7037cfa1c1715618308c8ff8e51

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,51 @@
+                      class="uploaded-file-chip"
+                      color="primary" selected
+                      (removed)="undoUploadedFile()">{{uploadedFile.name}} (size: {{uploadedFile.size}})
+                <mat-icon matChipRemove *ngIf="!readonly" #undoUploadBtn>cancel</mat-icon>
```
We can actually tell the server to delete the upload if they press the cancel button.
The way to do this is to PATCH an answer with a null value.
So right now, after the client uploads the contents of the file, your code correctly issues a PATCH that looks like:
```json
{"answers":[{"stableId":"FILE_UPLOAD_1","answerGuid":"SM27JVPJDB","value":"2kn3N8lkJXzmh12sFEMj"}]}
```
To delete the upload (i.e., when the matChipRemove is clicked) we would want to PATCH:
```json
{"answers":[{"stableId":"FILE_UPLOAD_1","answerGuid":"SM27JVPJDB","value":null}]}

alextanlan replied:
Right. It should work exactly like you wrote.
For example:
1.Upload a file
![upload1](https://user-images.githubusercontent.com/7396837/119811332-a767f000-beef-11eb-9187-9c0a08d8ba88.png)

2. Deleting the already uploaded file
![upload2](https://user-images.githubusercontent.com/7396837/119811395-b8186600-beef-11eb-8619-00c4b0f9b7e8.png)

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -1,88 +1,212 @@
+        }
+
+        if (this.uploadedFile) {
+            this.openReuploadConfirmDualog();
```
Typo.

alextanlan replied:
Fixed at 4ccdfdfaeaaf697e67cfe838da8109dc0fda5462.
Thanks!

alextanlan committed:
Fix a typo

alextanlan committed:
Change label

alextanlan committed:
Fix hard-coded values

alextanlan committed:
Rename modalService to modalDialogService

alextanlan committed:
Fix translation messages

alextanlan committed:
Fix code-review comments

  - format file size to Mb
  - format mime-types to file extensions
  - fix tests

alextanlan committed:
Refactoring as to get rid of selectedFile field

alextanlan committed:
Add i18n for fileAnswer component

alextanlan committed:
Merge branch 'develop' into DDP-5806-file-upload-extended

# Conflicts:
#	ddp-workspace/projects/ddp-sdk/src/lib/ddp.module.ts

aglushko-broad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/confirmDialog/confirmDialog.component.html:
```diff
@@ -0,0 +1,16 @@
+    </span>
+    <div class="confirm-dialog-buttons">
+        <button mat-flat-button class="ctrl-btn confirm-btn"
+                color="warn"
```
are we sure that confirm button should be in red color?

alextanlan replied:
I am not. 
However, the dialogs for now notify about **_Remove_**/**_Re-upload_** a file. 
They are needed to highlight that the operation can not be undo. 

![image](https://user-images.githubusercontent.com/7396837/120309167-9d723280-c2dd-11eb-9dfe-5e0c0a4150cd.png)

![image](https://user-images.githubusercontent.com/7396837/120309236-ad8a1200-c2dd-11eb-8a91-c422a1792c14.png)

alextanlan replied:
Do you have some suggestions about it ?

aglushko-broad replied:
well, the color might be configurable I guess, but it might be extended later if we feel that the red color is the right one here

alextanlan replied:
@aglushko-broad, which color are you suggesting to have there by default ?

alextanlan replied:
Another thing is that the confirmation dialog is used in other places (i.g. modalActivityBlock component)

aglushko-broad replied:
I don't have any suggestion, my comment initially was rather about "should we just copy it from original dialog or maybe even discuss with the designer what's better here" but as long as I'm not a designer I can't say which color might be better here. As I already wrote in the previous comment if red sounds right for you feel free to do nothing and ignore the comment. It wouldn't be a problem to add config functionality in the future if we need it.

alextanlan replied:
Ok, then I propose leaving it as is for now because we do not have any comprehension how it will be used later.

alextanlan replied:
Made the button color configurable and set the default 'primary' theme color palette at ec6bcb0eb8955cf5f4c58a4b5f1c314c9b9c1448 
![image](https://user-images.githubusercontent.com/7396837/120460231-9a407a80-c3a1-11eb-8e11-812688b1e4dc.png)

Eventually, we can override the color in specific study css if needed.

aglushko-broad commented:
I'm wondering if we can upload a file automatically after the user selected the file? Why do we need to additionally ask her/him to click the upload button?

alextanlan commented:
> I'm wondering if we can upload a file automatically after the user selected the file? Why do we need to additionally ask her/him 
> to click the upload button?

1. There are no requirements/design mockups for it. Perhaps, we will need to change it when it is used in a real study. 
2. In my opinion, having this interim step a user can have a chance to
       - change their mind if a wrong file is selected by mistake
       - see the file name when a validation error occurs as to easily analyze the error

MocanaAtBroad commented on ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:
```diff
@@ -699,6 +699,22 @@
+            "ChooseFile": "Choose file",
+            "Uploaded": "Uploaded:",
+            "Size": "(size: {{size}})",
+            "MaxSizeErrorMessage": "File size exceeds allowed size",
```
Please reword to: "Maximum file size exceeded" (avoids using the word "size" twice).

alextanlan replied:
Fixed at fadf33db0ccdfa4570a290bc9f84d4fd41f8ddac

MocanaAtBroad commented on ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:
```diff
@@ -699,6 +699,22 @@
+            "Uploaded": "Uploaded:",
+            "Size": "(size: {{size}})",
+            "MaxSizeErrorMessage": "File size exceeds allowed size",
+            "InvalidTypeErrorMessage": "File type not belongs to allowed list"
```
"Type of file selected is not valid"

alextanlan replied:
Fixed at fadf33db0ccdfa4570a290bc9f84d4fd41f8ddac

MocanaAtBroad reviewed (commented):
Still looking...

MocanaAtBroad commented on ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:
```diff
@@ -699,6 +699,22 @@
+            "UndoBtnText": "OK",
+            "CancelBtnText": "Cancel",
+            "MaxSize": "Max size: {{maxSize}}",
+            "AllowedFiles": "Allowed files: {{allowedFileTypes}}",
```
`AllowedFileTypes` probably more descriptive.

alextanlan replied:
Fixed at fadf33db0ccdfa4570a290bc9f84d4fd41f8ddac

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/models/fileSizeMeasure.ts:
```diff
@@ -0,0 +1,4 @@
+export enum FILE_SIZE_MEASURE {
+    kB = 1024,
```
In my Mac, I see file size listed using "KB" and "MB" so better if we go with these abbreviations.

alextanlan replied:
Fixed at cc816fd0f9f2c1028aa89e71a22ca29403ea42db

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -1,45 +1,72 @@
+    fileToUpload: UploadFile | null;
+    uploadedFile: ActivityFileAnswerDto | null;
+    fileMaxSize: number;
+    allowedFileTypes: string[];
```
Consider getting rid of instance variables. fileMaxSize and allowedFileTypes come directly from `block`; just copies.

alextanlan replied:
Ok. Fixed at 1bfa63d0862c3e41cd28bbde460c29108f6d47de

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/models/uploadFile.ts:
```diff
@@ -0,0 +1,6 @@
+
+export interface UploadFile extends FileUploadResponse {
+    file: File;
+    isReadyToUpload: boolean;
```
Not finding how property `isReadyToUpload` is being used; only see places where the value is set to `true` or `false` but don't see anywhere it is read. Could you double-check `isReadytoUpload` is needed?

alextanlan replied:
Thanks for noticing.
Please see https://github.com/broadinstitute/ddp-angular/pull/581#discussion_r643471373
Fixed at b836160f17b9a3addca82f9f11b9faa17a2408b7

alextanlan replied:
Eventually removed (see https://github.com/broadinstitute/ddp-angular/pull/581#issuecomment-852943813)

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -52,37 +79,127 @@ export class ActivityFileAnswer implements OnDestroy {
                 .pipe(
```
Consider refactoring the method signature for `getUploadUrl` . This component should not care what the request body for REST API looks like.  Might be better keeping this structural knowledge out of component and include fileName, fileSize, etc as arguments in `getUploadUrl` rather than `FileUploadBody`

alextanlan replied:
Fixed at c5dca4a0a0de154ddb2a0f45182c503a4294e7b0

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -52,37 +79,127 @@ export class ActivityFileAnswer implements OnDestroy {
+        this.fileUploadService.uploadFile(this.fileToUpload.uploadUrl, this.fileToUpload.file)
             .pipe(
                 catchError(err => {
                     this.logger.logDebug('ActivityFileAnswer uploadFile error:', err);
```
This line gives me Typescript compiler error. Looks to me like we need to set the method signature in the LoggingService correctly (currently expects 0 arguments which is obviously incorrect).

alextanlan replied:
It is interesting taking into account that `logger.logDebug` is called similarly in 26 places (except that one). 
Can you please provide a screenshot with Typescript compiler error ? 
Perhaps we just need to add `tslint:disable` there ?

MocanaAtBroad replied:
Not something you introduced in your PR anyways.
Fix  in this PR https://github.com/broadinstitute/ddp-angular/pull/595/files

alextanlan replied:
Yes, I saw your fix. 
It's odd that we didn't see the error earlier. 
Another point is that `console.debug` should have a list of arguments "out-of-the-box" 
(https://console.spec.whatwg.org/#debug)

MocanaAtBroad replied:
Just copied the console.debug method signature (I think the error has been there all along but it works like a "warning") the correct Javascript is generated anyways.

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/pipes/fileSizeFormatter.pipe.ts:
```diff
@@ -0,0 +1,14 @@
+        const measure = isFileSizeLessThenOneMb ? FILE_SIZE_MEASURE.kB : FILE_SIZE_MEASURE.Mb;
+        return (value / measure).toFixed(fractionDigits) + FILE_SIZE_MEASURE[measure];
+    }
+}
```
The pipe is a good idea!

MocanaAtBroad replied:
We should deal with files that are less than 1KB in size (in our test configuration with have a file with max size of 100 bytes).
![Screen Shot 2021-06-01 at 1 10 44 PM](https://user-images.githubusercontent.com/29984380/120364438-3d7a8c80-c2db-11eb-8016-7403985e2b7b.png)

MocanaAtBroad replied:
I think there are a lot of threshold and corner cases here in when to use bytes, MBs decimals,etc.
I am seeing that this is problem that other people have tried to solve. Maybe we should just use the this library or copy the function?
https://www.npmjs.com/package/filesize

alextanlan replied:
Improved at b05a866d623b7e42d7aee4ea49f566984c9507e6

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -52,37 +79,127 @@ export class ActivityFileAnswer implements OnDestroy {
                 .pipe(
                     catchError(err => {
                         this.logger.logDebug('ActivityFileAnswer getUploadUrl error:', err);
+                        this.errorMessage = err.message;
```
We should clear `errorMessage` whenever a new file is selected
1. We select a file that is too big
2. UI shows error "File is too big" (or something like that)
3. We select a file that is of allowed size.
4. UI keeps showing "File is too big" error.

alextanlan replied:
My bad. 
Fixed at 8d677f21d8cc286f7d0db97f4f7ce19197a03dcf

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,58 @@
 
+        <div class="selected-file">{{fileToUpload?.file.name}}</div>
+
+        <button *ngIf="!readonly" #uploadBtn
```
We should disable or hide "upload" button if there is no file selected.

alextanlan replied:
Right. I lost it by mistake during a previous refactoring.
Returned back at b836160f17b9a3addca82f9f11b9faa17a2408b7.

alextanlan replied:
Eventually removed (see https://github.com/broadinstitute/ddp-angular/pull/581#issuecomment-852943813)

MocanaAtBroad commented:
>> alextanlan 5 days ago • 
 Author Member
As for i18n:
I have added i18n for ddp-angio only (in test purpose) at 96f9087

>> we don't know for now where the File Upload functionality will be applied (in which activity).
It does not make sense to apply FileUpload translations for all existing studies right now, in my opinion.

>> What do you think about it?

The way we currently do translations, we copy the translations entries from SDK to the project that is using SDK.
In this case, because feature is not being used at this time in any project, don't see reason to copy new entries. For new projects (even if they don't have file upload), I think OK for stuff to be copied.

But bottom line: no need to copy to translations to any existing study at this time.

MocanaAtBroad commented:
>> aglushko-broad 6 hours ago Member
I don't have any suggestion, my comment initially was rather about "should we just copy it from original dialog or maybe even discuss with the designer what's better here" but as long as I'm not a designer I can't say which color might be better here. As I already wrote in the previous comment if red sounds right for you feel free to do nothing and ignore the comment. It wouldn't be a problem to add config functionality in the future if we need it.

>> @alextanlan alextanlan 6 hours ago • 
 Author Member
Ok, then I propose leaving it as is for now because we do not have any comprehension how it will be used later.

I would not choose red for the button and use whatever we are using for confirmation elsewhere. Gray? Green?

But would add that at this time we don't really have a web design we are working out of and might have to revisit a few things, like the overall layout, fonts, etc. 
Would be nice to combine the button that opens the file selection dialog with the file drag and drop area; have one thing that you can click or drag files into instead of separate widgets.
But I think this is OK for now. Looking for the basic functionality to be in place and we can come back and revisit when we have a clean design.

MocanaAtBroad reviewed (commented):
> I'm wondering if we can upload a file automatically after the user selected the file? Why do we need to additionally ask her/him to click the upload button?

Would be good to get rid of additional step/need for button (I have MANY times personally thought I had uploaded attachments to later find out that I forgot to click the upload button)

alextanlan committed:
Fix messages after code-review

alextanlan committed:
Fix file size titles

alextanlan committed:
Return back missed attribute

alextanlan committed:
Get rid of excess variables

alextanlan committed:
Clear error message before file selection

alextanlan committed:
Fix after code-review

alextanlan committed:
Customize confirm button color (in confirm dialog)

alextanlan committed:
Combine file select button and drag-and-drop block

alextanlan commented:
> Would be nice to combine the button that opens the file selection dialog with the file drag and drop area; have one thing that you can click or drag files into instead of separate widgets.

done at 56662b7502b58c3e4a7e0038d73d34c11fdeb282

alextanlan committed:
Get rid of Upload/Reupload button (as an additional step during uploading)

alextanlan commented:
>Would be good to get rid of additional step/need for button (I have MANY times personally thought I had uploaded attachments to later find out that I forgot to click the upload button)

Removed Upload/Reupload button (as an additional step during the file uploading) at fd88a6eb17817ca6762f1b7201ebb5466eb0f148

alextanlan committed:
Improve file size formatter

alextanlan committed:
Merge branch 'develop' into DDP-5806-file-upload-extended

MocanaAtBroad commented:
We need to disable the submit button while we get the URL to do the authorization and while there is an upload in progress. We might be able to use the component busy mechanism we use from mailing address? Take a look.
(You probably know this already, but the trick I use to simulate a slow connection/longer time for upload is to turn on throttling in the Chrome Dev tools. Easier to see the submit button enabling/disabling)
![Screen Shot 2021-06-03 at 9 04 56 AM](https://user-images.githubusercontent.com/29984380/120649586-ec89a600-c44a-11eb-8b02-5ad8d194b01a.png)

MocanaAtBroad commented:
I think we also need some kind of indicator that there is an upload in progress. Perhaps just a simple GIF? Perhaps some text saying something like "Uploading MyFile.txt..."
We don't have nice design to work from, but hopefully at some point we will and we can just swap out the images, tweak layouts, etc., but the basic elements and functionality will be there.

alextanlan commented:
> Was this implementation copied more or less from the filesize library I linked to?

Yes, it was. The thing is that the library has 'BSD-3-Clause' License. I was not sure if we can use it.
Therefore I extracted/copied the main idea of the transform method.

alextanlan committed:
Disable the activity submit button while file upload requests are in progress.

alextanlan committed:
Add progress bar for file uploading

alextanlan commented:
> We need to disable the submit button while we get the URL to do the authorization and while there is an upload in progress. We might be able to use the component busy mechanism we use from mailing address? Take a look.

Implemented at 44fcefbe583704af9e3fd2f28cdd3063a3f731bc

alextanlan commented:
> I think we also need some kind of indicator that there is an upload in progress. Perhaps just a simple GIF? Perhaps some text saying something like "Uploading MyFile.txt..."


Added a progres-bar at 3f38c8b403dae066497a6459995bc2975340a0e3

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.ts:
```diff
@@ -1,88 +1,203 @@
+        this.initUploadedFile();
     }
 
     onFilesSelected(files: File[]): void {
```
I am getting a type warning in template. Change below gets rid of it.
```suggestion
    onFilesSelected(files: File[] | FileList): void {
```

alextanlan replied:
Fixed at f929b71f78f178edac82e717e16d6cc391afd774

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-file-answer/activityFileAnswer.component.html:
```diff
@@ -1,27 +1,46 @@
-        <span class="file-name">{{file?.name || 'No file uploaded'}}</span>
+
+    <div class="uploaded-file">
+        <div class="title" translate>SDK.FileUpload.Uploaded</div>
```
I don't think this label makes sense anymore. Even when it is true that it is "**Uploaded**", it is already clear from the mat-chip"
![Screen Shot 2021-06-06 at 4 39 28 PM](https://user-images.githubusercontent.com/29984380/120939430-0c88c600-c6e6-11eb-97dc-e8212cc2075f.png)

Consider changing it to appear only when upload in progress and with the name of the file.

alextanlan replied:
Changed at 8f3c18efd56ba1cbea0060bae1c156be4ffb7d7f

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:
```diff
@@ -58,7 +58,8 @@ import { SubmissionManager } from '../../services/serviceAgents/submissionManage
                                            [studyGuid]="studyGuid"
-                                           [activityGuid]="activityGuid">
+                                           [activityGuid]="activityGuid"
+                                           (componentBusy)="embeddedComponentBusy.emit($event)">
```
I think this is the correct thing to do (emit true or false to indicate whether busy or not), but now the @Output() variable name is incorrect. We should probably rename now just `componentBusy` as it is just not embedded components that we are worried about.

alextanlan replied:
Renamed at e76e653f7aac823ccc154ca39530a64606c8a2cd

MocanaAtBroad reviewed (commented):
Now that you have added the "Upload in progress" indicator, seems to me that the last piece of functionality missing here is to be able to cancel the upload. A "cancel" button with an "Are you sure you want to cancel upload?" confirmation dialog. Hard to add?

alextanlan committed:
Fix type

alextanlan committed:
Rename output emitter for activitySection.component

alextanlan committed:
Code improvement

  The less instance variables we have,
  the less state to track
  and less chance of subtle bugs coming into play.

alextanlan committed:
Fix tests

alextanlan commented:
> After last refactor and reviewing latest changes, I am under impression that this interface can just go away. Only thing in it that matters is the File (one attribute) and the upload url is the other argument to actually do the upload and appears it would be easier to just pass it as a function argument. MIGHT be able to get rid of fileToUpload instance variable altogether, but not 100% sure, but please do take a look. The less instance variables we have, the less state to track and less chance of subtle bugs coming into play.

Changed at 8f3c18efd56ba1cbea0060bae1c156be4ffb7d7f

alextanlan commented:
> Now that you have added the "Upload in progress" indicator, seems to me that the last piece of functionality missing here is to be able to cancel the upload. A "cancel" button with an "Are you sure you want to cancel upload?" confirmation dialog. Hard to add?

It seems to be tricky for me. The thing is that we patch answers "in background" and can not control from a component if the answer has been already patched.

MocanaAtBroad commented:
> > Now that you have added the "Upload in progress" indicator, seems to me that the last piece of functionality missing here is to be able to cancel the upload. A "cancel" button with an "Are you sure you want to cancel upload?" confirmation dialog. Hard to add?
> 
> It seems to be tricky for me. The thing is that we patch answers "in background" and can not control from a component if the answer has been already patched.

Well, I think the Cancel button would only be visible during the "Uploading" stage and disappear as soon as upload is completed/before the PATCH is queued on the client. Once upload is completed , it should be no more than a few seconds before they are able to click to remove the upload.

The case I am worried about is when we have really large files (e.g.., 100MB on a bad connection) and realize they uploaded wrong file. Seems like punishment to make them wait until upload is completed before they can do the real one.

alextanlan committed:
Add upload canceling

alextanlan committed:
Improve i18n for file upload

alextanlan commented:
> I think the Cancel button would only be visible during the "Uploading" stage and disappear as soon as upload is completed/before the PATCH is queued on the client.

Thank you for the idea ! Applied at d11f84322337471b79b209ef1547cc905999a58c

alextanlan committed:
Merge branch 'develop' into DDP-5806-file-upload-extended

MocanaAtBroad commented on ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts:
```diff
@@ -60,9 +60,16 @@ export class EmbeddedActivityBlockComponent implements OnInit {
 
-        const dialogRef = this.dialog.open(ActivityDeleteDialogComponent, config);
+        const dialogRef = this.dialog.open(ConfirmDialogComponent, config);
         dialogRef.afterClosed().subscribe((confirmDelete: boolean) => {
```
Do we have unsubscribe from dialog events? (Highlighting this instance but there are a couple other ones in file)

aglushko-broad replied:
we don't need to unsubscribe from afterClosed result since the observable itself completes

MocanaAtBroad reviewed (commented):
The cancellation looks and works nice!
Just had one question observation.
